### PR TITLE
Plugins

### PR DESCRIPTION
**Pull request by denCalderon**
Add functionality to handle third party plugins

### CODE REVIEW

File: kit/commands/init.py [hunk 1]
```diff
@@ -137,8 +147,10 @@ def init_hekit(args) -> None:
     path_line = "PATH=$HEKITPATH:$PATH\n"
     # 2-Register hekit link and hekit.py script to enable tab completion
     eval_lines = (
-        "if [ -n $(type -p register-python-argcomplete) ]; then\n"
-        ' eval "$(register-python-argcomplete hekit)"\n'
+        "if [ -n $(type -p register-python-argcomplete3) ]; then\n"
```
Comment (hamishun) [inline kit/commands/init.py:150]: Not sure if this is a good idea. I think this is for the argcomplete that is installed by apt.

File: kit/hekit.py [hunk 2]
```diff
@@ -43,12 +43,8 @@ def parse_cmdline() -> Tuple:
...
     # create subparsers for each command
     subparsers = parser.add_subparsers(help="sub-command help")
-
-    for func in discover_subparsers_from(["commands", "tools"], hekit_root_dir / "kit"):
-        try:
-            func(subparsers)
-        except TypeError:
-            func(subparsers, hekit_root_dir)
+    register_subparser(subparsers, ["commands", "tools"], hekit_root_dir / "kit")
+    register_subparser(subparsers, get_plugins_by_state(), PluginsConfig.ROOT_DIR)
```
Comment (hamishun) [inline kit/hekit.py:47]: I don't think it is a good idea to have `get_plugins_by_state` have a default state here. It is not obvious reading the code here that the default is enabled. Just be explicit for clarity.

**Review by hamishun — CHANGES_REQUESTED** [review]

File: tests/test_util_files.py [hunk 3]
```diff
@@ -61,10 +69,32 @@ def test_list_dirs(get_toolkit_path):
...
+
+
+def test_load_toml(get_toolkit_path):
+    file = f"{get_toolkit_path}/tests/input_files/default.config"
+    act_dict = load_toml(file)
+    assert "~/.hekit_test/components" == act_dict["repo_location"]
+
+
+def test_load_toml_symlink(get_toolkit_path):
+    file = f"{get_toolkit_path}/tests/input_files/default_symlink.config"
```
Comment (hamishun) [inline tests/test_util_files.py:84]:
```suggestion
    file = get_toolkit_path / "tests/input_files/default_symlink.config"
```

File: kit/commands/new.py [hunk 4]
```diff
@@ -79,15 +79,15 @@ def create_directory_structure(project_name: str, project_path: Path) -> None:
 def create_new_project(args) -> None:
     """create a new project"""
     project_name = args.name
-    project_path = Path(args.directory).resolve().expanduser()
+    project_path = Path(args.directory).expanduser().resolve()
     project_path = project_path / "projects" / f"{project_name}"
```
Comment (hamishun) [inline kit/commands/new.py:83]:
```suggestion
    project_path = project_path / "projects" / project_name
```

**Review by hamishun — CHANGES_REQUESTED** [review]

**Review by hamishun — DISMISSED** [review]

File: kit/utils/files.py [hunk 5]
```diff
@@ -40,3 +47,22 @@ def create_default_workspace(dir_path: str = "~") -> Path:
     workspace_path = Path(dir_path).expanduser() / ".hekit"
     workspace_path.mkdir(exist_ok=True)
     return workspace_path
+
+
+def load_toml(file_name: PathType) -> dict:
+    """Load a toml file and returns its content as a dict"""
```
Comment (jlhcrawford) [inline kit/utils/files.py:53]:
```suggestion
    """Load a toml file and return its content as a dict"""
```

File: tests/test_cmd_plugins.py [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+    """Verify that the SW raises an error
+    when the plugin file is not found"""
+    plugin_name = "MyPlugin"
+    with pytest.raises(FileNotFoundError) as exc_info:
+        get_plugin_type(Path(plugin_name))
+    assert str(exc_info.value) == f"File '{plugin_name}' not found"
+
+
+def test_get_plugin_type_file_not_supported():
+    """Verify that the SW return a dict
```
Comment (jlhcrawford) [inline tests/test_cmd_plugins.py:67]:
```suggestion
    """Verify that the SW returns a dict
```

File: tests/test_cmd_plugins.py [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+    that was read from a toml file"""
+    tests_path = Path(__file__).resolve().parent
+    file = tests_path / "input_files/default.config"
+    with pytest.raises(TypeError) as exc_info:
+        get_plugin_type(file)
+    assert str(exc_info.value) == "This program only supports tarball or zip files"
+
+
+def test_get_plugin_type_dir(tmp_path):
+    """Verify that the SW returns the correct plugin type"""
```
Comment (jlhcrawford) [inline tests/test_cmd_plugins.py:77]:
```suggestion
    """Verify that the SW returns the correct plugin type (directory)"""
```

File: tests/test_cmd_plugins.py [hunk 8]
```diff
@@ -0,0 +1,516 @@
...
+    plugin_path = create_plugins_files("test", tmp_path, PluginType.DIR)
+    assert PluginType.DIR == get_plugin_type(plugin_path)
+
+
+def test_get_plugin_type_tar(tmp_path):
+    """Verify that the SW returns the correct plugin type"""
```
Comment (jlhcrawford) [inline tests/test_cmd_plugins.py:83]:
```suggestion
    """Verify that the SW returns the correct plugin type (tarball)"""
```

File: tests/test_cmd_plugins.py [hunk 9]
```diff
@@ -0,0 +1,516 @@
...
+    plugin_path = create_plugins_files("test", tmp_path, PluginType.TAR)
+    assert PluginType.TAR == get_plugin_type(plugin_path)
+
+
+def test_get_plugin_type_zip(tmp_path):
+    """Verify that the SW returns the correct plugin type"""
```
Comment (jlhcrawford) [inline tests/test_cmd_plugins.py:89]:
```suggestion
    """Verify that the SW returns the correct plugin type (zip file)"""
```

File: tests/test_cmd_plugins.py [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+    mockers = Mockers(mocker)
+
+    update_plugin_state(args)
+    mockers.mock_dump_toml.assert_not_called()
+    mockers.mock_print.assert_called_with(
+        f"Plugin {args.plugin} is already enabled in the system"
+    )
+
+
+def test_update_plugin_state_already_disabled(mocker):
```
Comment (jlhcrawford) [inline tests/test_cmd_plugins.py:385]: This test does not test for a disabled plugin but an enabled one like the test above.

File: tests/test_cmd_plugins.py [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+    args = MockArgs()
+    args.all = True
+    mockers = Mockers(mocker)
+    mock_input = mocker.patch("kit.commands.plugin.input", return_value="y")
+
+    remove_plugin(args)
+    mock_input.assert_called_once()
+    assert 2 == mockers.mock_remove_dir.call_count
+    mockers.mock_dump_toml.assert_called_once()
+    mockers.mock_print.assert_called_with("All Plugins were uninstalled successfully")
```
Comment (jlhcrawford) [inline tests/test_cmd_plugins.py:353]:
```suggestion
    mockers.mock_print.assert_called_with("All plugins were uninstalled successfully")
```

File: kit/commands/plugin.py [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+            "All plugins will be deleted. Do you want to continue? (y/n) "
+        )
+        if user_answer not in ("y", "Y"):
+            return
+
+        for plugin_name in plugin_dict.keys():
+            remove_plugin_dir(plugin_name)
+
+        plugin_dict.clear()
+        print("All Plugins were uninstalled successfully")
```
Comment (jlhcrawford) [inline kit/commands/plugin.py:161]:
```suggestion
        print("All plugins were uninstalled successfully")
```

**Review by jlhcrawford — CHANGES_REQUESTED** [review]
Some typos but there is a duplicate plugin pytest that needs changing.

**Review by jlhcrawford — APPROVED** [review]

**Pull request merged**